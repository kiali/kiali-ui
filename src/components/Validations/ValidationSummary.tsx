import * as React from 'react';
import { CSSProperties } from 'react';
import { ValidationTypes } from '../../types/IstioObjects';
import { style } from 'typestyle';
import { Text, TextVariants, Tooltip, TooltipPosition } from '@patternfly/react-core';
import Validation from './Validation';

interface Props {
  id: string;
  errors: number;
  warnings: number;
  objectCount?: number;
  style?: CSSProperties;
}

const tooltipListStyle = style({
  textAlign: 'left',
  border: 0,
  padding: '0 0 0 0',
  margin: '0 0 0 0'
});

const tooltipSentenceStyle = style({
  textAlign: 'center',
  border: 0,
  padding: '0 0 0 0',
  margin: '0 0 0 0'
});

export class ValidationSummary extends React.PureComponent<Props> {
  getTypeMessage = (count: number, type: ValidationTypes): string => {
    return count > 1 ? `${count} ${type}s found` : `${count} ${type} found`;
  };

  severitySummary() {
    const issuesMessages: string[] = [];

    if (this.props.errors > 0) {
      issuesMessages.push(this.getTypeMessage(this.props.errors, ValidationTypes.Error));
    }

    if (this.props.warnings > 0) {
      issuesMessages.push(this.getTypeMessage(this.props.warnings, ValidationTypes.Warning));
    }

    if (issuesMessages.length === 0) {
      issuesMessages.push('No issues found');
    }

    return issuesMessages;
  }

  severity() {
    let severity = ValidationTypes.Correct;
    if (this.props.errors > 0) {
      severity = ValidationTypes.Error;
    } else if (this.props.warnings > 0) {
      severity = ValidationTypes.Warning;
    }

    return severity;
  }

  tooltipNA() {
    return <Text className={tooltipSentenceStyle}>No Istio config objects found</Text>;
  }

  tooltipNoValidationAvailable() {
    return <Text className={tooltipListStyle}>No Istio config validation available</Text>;
  }

  tooltipSummary() {
    return (
      <>
        <Text style={{ textAlign: 'left', textEmphasis: 'strong' }} component={TextVariants.p}>
          Istio config objects analyzed: {this.props.objectCount}
        </Text>
        <div className={tooltipListStyle}>
          {this.severitySummary().map(cat => (
            <div key={cat}>{cat}</div>
          ))}
        </div>
      </>
    );
  }

  tooltipContent() {
    if (this.props.objectCount !== undefined) {
      if (this.props.objectCount === 0) {
        return this.tooltipNA();
      } else {
        return this.tooltipSummary();
      }
    } else {
      return this.tooltipNoValidationAvailable();
    }
  }

  tooltipBase() {
    return this.props.objectCount === undefined || this.props.objectCount > 0 ? (
      <Validation iconStyle={this.props.style} severity={this.severity()} />
    ) : (
      <div style={{ display: 'inline-block', marginLeft: '5px' }}>N/A</div>
    );
  }

  render() {
    return (
      <Tooltip
        aria-label={'Validations list'}
        position={TooltipPosition.auto}
        enableFlip={true}
        content={this.tooltipContent()}
      >
        {this.tooltipBase()}
      </Tooltip>
    );
  }
}

export default ValidationSummary;
