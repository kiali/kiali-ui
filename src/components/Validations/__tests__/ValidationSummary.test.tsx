import * as React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import { shallow } from 'enzyme';
import ValidationSummary from '../ValidationSummary';

const testScenario = (summary: any) => {
  const wrapper = shallow(summary);
  expect(shallowToJson(wrapper)).toBeDefined();
  expect(shallowToJson(wrapper)).toMatchSnapshot();
};

describe('ValidationSummary renders', () => {
  it('success icon when all components are valid', () => {
    testScenario(<ValidationSummary id={'1'} errors={0} warnings={0} objectCount={1} />);
    testScenario(<ValidationSummary id={'1'} errors={0} warnings={0} objectCount={4} />);
  });

  it('warning icon when all components are valid', () => {
    testScenario(<ValidationSummary id={'2'} errors={0} warnings={1} objectCount={1} />);
    testScenario(<ValidationSummary id={'2'} errors={0} warnings={1} objectCount={3} />);
  });

  it('error icon when all components are valid', () => {
    testScenario(<ValidationSummary id={'3'} errors={1} warnings={0} objectCount={1} />);
    testScenario(<ValidationSummary id={'3'} errors={1} warnings={2} objectCount={3} />);
  });

  it('N/A when all components are valid', () => {
    testScenario(<ValidationSummary id={'4'} errors={0} warnings={0} objectCount={0} />);
  });
});
