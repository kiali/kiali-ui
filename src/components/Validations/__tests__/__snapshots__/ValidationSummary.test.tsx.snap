// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationSummary renders N/A when all components are valid 1`] = `
<Tooltip
  appendTo={[Function]}
  aria="describedby"
  aria-label="Validations list"
  boundary="window"
  className=""
  content={
    <Text
      className="f12qq397"
    >
      No Istio configuration objects found in this namespace
    </Text>
  }
  distance={15}
  enableFlip={true}
  entryDelay={500}
  exitDelay={500}
  flipBehavior={
    Array [
      "top",
      "right",
      "bottom",
      "left",
      "top",
      "right",
      "bottom",
    ]
  }
  id=""
  isAppLauncher={false}
  isContentLeftAligned={false}
  isVisible={false}
  maxWidth="18.75rem"
  position="auto"
  tippyProps={Object {}}
  trigger="mouseenter focus"
  zIndex={9999}
>
  <small
    style={
      Object {
        "fontSize": "65%",
        "marginLeft": "5px",
      }
    }
  >
    N/A
  </small>
</Tooltip>
`;

exports[`ValidationSummary renders error icon when all components are valid 1`] = `
<Tooltip
  appendTo={[Function]}
  aria="describedby"
  aria-label="Validations list"
  boundary="window"
  className=""
  content={
    <React.Fragment>
      <Text
        component="h4"
        style={
          Object {
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        Istio Config Validation
      </Text>
      <Text
        component="h5"
        style={
          Object {
            "paddingLeft": "1em",
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        1 object analyzed
      </Text>
      <div
        className="flhywg2"
      >
        <div>
          1 error found
        </div>
      </div>
    </React.Fragment>
  }
  distance={15}
  enableFlip={true}
  entryDelay={500}
  exitDelay={500}
  flipBehavior={
    Array [
      "top",
      "right",
      "bottom",
      "left",
      "top",
      "right",
      "bottom",
    ]
  }
  id=""
  isAppLauncher={false}
  isContentLeftAligned={false}
  isVisible={false}
  maxWidth="18.75rem"
  position="auto"
  tippyProps={Object {}}
  trigger="mouseenter focus"
  zIndex={9999}
>
  <Validation
    severity="error"
  />
</Tooltip>
`;

exports[`ValidationSummary renders error icon when all components are valid 2`] = `
<Tooltip
  appendTo={[Function]}
  aria="describedby"
  aria-label="Validations list"
  boundary="window"
  className=""
  content={
    <React.Fragment>
      <Text
        component="h4"
        style={
          Object {
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        Istio Config Validation
      </Text>
      <Text
        component="h5"
        style={
          Object {
            "paddingLeft": "1em",
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        3 objects analyzed
      </Text>
      <div
        className="flhywg2"
      >
        <div>
          1 error found
        </div>
        <div>
          2 warnings found
        </div>
      </div>
    </React.Fragment>
  }
  distance={15}
  enableFlip={true}
  entryDelay={500}
  exitDelay={500}
  flipBehavior={
    Array [
      "top",
      "right",
      "bottom",
      "left",
      "top",
      "right",
      "bottom",
    ]
  }
  id=""
  isAppLauncher={false}
  isContentLeftAligned={false}
  isVisible={false}
  maxWidth="18.75rem"
  position="auto"
  tippyProps={Object {}}
  trigger="mouseenter focus"
  zIndex={9999}
>
  <Validation
    severity="error"
  />
</Tooltip>
`;

exports[`ValidationSummary renders success icon when all components are valid 1`] = `
<Tooltip
  appendTo={[Function]}
  aria="describedby"
  aria-label="Validations list"
  boundary="window"
  className=""
  content={
    <React.Fragment>
      <Text
        component="h4"
        style={
          Object {
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        Istio Config Validation
      </Text>
      <Text
        component="h5"
        style={
          Object {
            "paddingLeft": "1em",
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        1 object analyzed
      </Text>
      <div
        className="flhywg2"
      >
        <div>
          No issues found
        </div>
      </div>
    </React.Fragment>
  }
  distance={15}
  enableFlip={true}
  entryDelay={500}
  exitDelay={500}
  flipBehavior={
    Array [
      "top",
      "right",
      "bottom",
      "left",
      "top",
      "right",
      "bottom",
    ]
  }
  id=""
  isAppLauncher={false}
  isContentLeftAligned={false}
  isVisible={false}
  maxWidth="18.75rem"
  position="auto"
  tippyProps={Object {}}
  trigger="mouseenter focus"
  zIndex={9999}
>
  <Validation
    severity="correct"
  />
</Tooltip>
`;

exports[`ValidationSummary renders success icon when all components are valid 2`] = `
<Tooltip
  appendTo={[Function]}
  aria="describedby"
  aria-label="Validations list"
  boundary="window"
  className=""
  content={
    <React.Fragment>
      <Text
        component="h4"
        style={
          Object {
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        Istio Config Validation
      </Text>
      <Text
        component="h5"
        style={
          Object {
            "paddingLeft": "1em",
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        4 objects analyzed
      </Text>
      <div
        className="flhywg2"
      >
        <div>
          No issues found
        </div>
      </div>
    </React.Fragment>
  }
  distance={15}
  enableFlip={true}
  entryDelay={500}
  exitDelay={500}
  flipBehavior={
    Array [
      "top",
      "right",
      "bottom",
      "left",
      "top",
      "right",
      "bottom",
    ]
  }
  id=""
  isAppLauncher={false}
  isContentLeftAligned={false}
  isVisible={false}
  maxWidth="18.75rem"
  position="auto"
  tippyProps={Object {}}
  trigger="mouseenter focus"
  zIndex={9999}
>
  <Validation
    severity="correct"
  />
</Tooltip>
`;

exports[`ValidationSummary renders warning icon when all components are valid 1`] = `
<Tooltip
  appendTo={[Function]}
  aria="describedby"
  aria-label="Validations list"
  boundary="window"
  className=""
  content={
    <React.Fragment>
      <Text
        component="h4"
        style={
          Object {
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        Istio Config Validation
      </Text>
      <Text
        component="h5"
        style={
          Object {
            "paddingLeft": "1em",
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        1 object analyzed
      </Text>
      <div
        className="flhywg2"
      >
        <div>
          1 warning found
        </div>
      </div>
    </React.Fragment>
  }
  distance={15}
  enableFlip={true}
  entryDelay={500}
  exitDelay={500}
  flipBehavior={
    Array [
      "top",
      "right",
      "bottom",
      "left",
      "top",
      "right",
      "bottom",
    ]
  }
  id=""
  isAppLauncher={false}
  isContentLeftAligned={false}
  isVisible={false}
  maxWidth="18.75rem"
  position="auto"
  tippyProps={Object {}}
  trigger="mouseenter focus"
  zIndex={9999}
>
  <Validation
    severity="warning"
  />
</Tooltip>
`;

exports[`ValidationSummary renders warning icon when all components are valid 2`] = `
<Tooltip
  appendTo={[Function]}
  aria="describedby"
  aria-label="Validations list"
  boundary="window"
  className=""
  content={
    <React.Fragment>
      <Text
        component="h4"
        style={
          Object {
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        Istio Config Validation
      </Text>
      <Text
        component="h5"
        style={
          Object {
            "paddingLeft": "1em",
            "textAlign": "left",
            "textEmphasis": "strong",
          }
        }
      >
        3 objects analyzed
      </Text>
      <div
        className="flhywg2"
      >
        <div>
          1 warning found
        </div>
      </div>
    </React.Fragment>
  }
  distance={15}
  enableFlip={true}
  entryDelay={500}
  exitDelay={500}
  flipBehavior={
    Array [
      "top",
      "right",
      "bottom",
      "left",
      "top",
      "right",
      "bottom",
    ]
  }
  id=""
  isAppLauncher={false}
  isContentLeftAligned={false}
  isVisible={false}
  maxWidth="18.75rem"
  position="auto"
  tippyProps={Object {}}
  trigger="mouseenter focus"
  zIndex={9999}
>
  <Validation
    severity="warning"
  />
</Tooltip>
`;
